libsubid/subid.h

```diff
@@ -13,5 +14,73 @@ enum subid_type {
 	ID_TYPE_GID = 2
 };
 
+enum subid_status {
+	SUBID_STATUS_SUCCESS = 0,
+	SUBID_STATUS_UNKNOWN_USER = 1,
+	SUBID_STATUS_ERROR_CONN = 2,
+	SUBID_STATUS_ERROR = 3,
+};
+
+struct subid_nss_ops {
+	/*
+	 * nss_has_any_range: does a user own any subid range
+	 *
+	 * @owner: username
+	 * @idtype: subuid or subgid
+	 * @result: true if a subid allocation was found for @owner
+	 *
+	 * returns success if the module was able to determine an answer (true or false),
+	 * else an error status.
+	 */
+	enum subid_status (*has_any_range)(const char *owner, enum subid_type idtype, bool *result);
+
+	/*
+	 * nss_has_range: does a user own a given subid range
+	 *
+	 * @owner: username
+	 * @start: first subid in queried range
+	 * @count: number of subids in queried range
+	 * @idtype: subuid or subgid
+	 * @result: true if @owner has been allocated the subid range.
+	 *
+	 * returns success if the module was able to determine an answer (true or false),
+	 * else an error status.
+	 */
+	enum subid_status (*has_range)(const char *owner, unsigned long start, unsigned long count, enum subid_type idtype, bool *result);
+
+	/*
+	 * nss_list_owner_ranges: list the subid ranges delegated to a user.
+	 *
+	 * @owner - string representing username being queried
+	 * @id_type - subuid or subgid
+	 * @ranges - * a NULL-terminated array of struct subordinate_range,
+	 * or NULL.  The returned array of struct subordinate_range must be
+	 * freed by the caller and it's members, if not NULL.
+	 * @ok - whether the result (even if 0) was successful from the nss module.
+
+	 * returns success if the module was able to determine an answer (true or false),
+	 * else an error status.
+	 */
+	enum subid_status (*list_owner_ranges)(const char *owner, enum subid_type id_type, struct subordinate_range ***ranges, bool *ok);
+
+	/*
+	 * nss_find_subid_owners: find uids who own a given subuid or subgid.
+	 *
+	 * @owner - string representing username being queried
+	 * @uids - pointer to an array of uids which will be allocated by
+	 *         nss_find_subid_owners()
+	 * @id_type - subuid or subgid
+	 * @count - number of uids found
+	 * @ok - whether the result (even if 0) was successful from the nss module.
+	 *
+	 * returns success if the module was able to determine an answer (true or false),
+	 * else an error status.
+	 */
+	enum subid_status (*find_subid_owners)(unsigned long id, uid_t **uids, enum subid_type id_type, int *count, bool *ok);
```

Comment:
Same as above about copy&paste and `ok` plus:
 - `id` isn't described
 - why do we use `count` here and NULL-terminated array in `list_owner_ranges()`? Wouldn't it make sense to have "symmetric" API?
 - could we have the same logical order of args - "subject - id_type - result" - as in `list_owner_ranges()`?

Something like:
`enum subid_status (*find_subid_owners)(unsigned long subid, enum subid_type id_type, uid_t **uids);`

Reply:
re count - yeah it is inconsistent, it has kinda bothered me.  Which one do people prefer, a returned count or ending NULL?

libsubid is already out there using the ending null, and i really don't want to have to have libsubid massage what it gets back from nss.

But as far as I know noone is yet using libsubid, so maybe it's still ok to change so far.

Reply:
> re count - yeah it is inconsistent, it has kinda bothered me. Which one do people prefer, a returned count or ending NULL?

Any option is fine for me. Just preferably to have it consistent.